cumulus_library/apis/umls.py

```diff
@@ -0,0 +1,152 @@
+        if "http:" in url:
+            url = url.replace("http:", "https:")
+        if "/res/" not in url:
+            url = url.replace("/fhir/", "/fhir/res/")
+        response = self.session.get(url)
+        if response.status_code == 404:
+            raise errors.ApiError(f"Url not found: {url}")
+        res_json = response.json()
+        if "compose" in res_json and "include" in res_json["compose"]:
+            nested_valuesets = []
+            for record in res_json["compose"]["include"]:
+                if "valueSet" in record:
+                    nested_get = self.get_vsac_valuesets(url=record["valueSet"][0])
+                    nested_valuesets.append(nested_get[0])
+
+            return [response.json(), *nested_valuesets]
+        return response.json()
+
+    def download_umls_files(
+        self,
+        target: str = "umls-metathesaurus-mrconso-file",
+        path: pathlib.Path | None = None,
+    ):
+        """Downloads an available file from the UMLS Download API and unzips it
+        target: the UMLS resource to download (default: the MRCONSO.RRF file)
+        path: the path on disk to write to
+
+        See https://documentation.uts.nlm.nih.gov/automating-downloads.html for more
+        info about the available downloads
+        """
+        if target not in VALID_UMLS_DOWNLOADS:
+            raise errors.ApiError(
+                f"'{target}' is not a valid umls download type.\n\n"
+                f"Expected values: {','.join(VALID_UMLS_DOWNLOADS)}"
+            )
+        if path is None:
+            path = pathlib.Path.cwd()
+        release_payload = {"releaseType": target, "current": "true"}
+        file_meta = self.session.get(
+            "https://uts-ws.nlm.nih.gov/releases", params=release_payload
+        ).json()[0]
+
+        # This particular endpoint requires the API key as a param rather than a
+        # basic auth header ¯\_(ツ)_/¯.
+        download_payload = {
+            "url": file_meta["downloadUrl"],
+            "apiKey": self.api_key,
+        }
+        download_res = requests.get(
+            "https://uts-ws.nlm.nih.gov/download", params=download_payload, stream=True
+        )
+
+        with open(path / file_meta["fileName"], "wb") as f:
+            chunks_read = 0
+            with base_utils.get_progress_bar() as progress:
+                task = progress.add_task(
+                    f"Downloading {file_meta['fileName']}", total=None
+                )
+                for chunk in download_res.iter_content(chunk_size=1024):
+                    f.write(chunk)
+                    chunks_read += 1
+                    progress.update(
+                        task,
+                        description=(
+                            f"Downloading {file_meta['fileName']}: "
+                            f"{chunks_read/1000} MB"
+                        ),
+                    )
+        base_utils.unzip_file(path / file_meta["fileName"], path)
```

Comment:
Does this process leave the zip file on disk? Maybe delete it after this? Or unzip as you stream, dunno how trivial the library makes that.

Reply:
it does leave the zip file on disk.

i could clean this up - i went back and forth on doing that, versus keeping the file around and then checking for it in the dir if it already existed and then skipping the download. this is clearly the worst of both worlds.

Reply:
ok, now this removes the zip file at the end.